dlt/destinations/impl/synapse/configuration.py

```diff
@@ -0,0 +1,106 @@
+from typing import Final, Any, List, Dict, Optional, ClassVar
+
+from dlt.common import logger
+from dlt.common.configuration import configspec
+from dlt.common.schema.typing import TTableIndexType, TSchemaTables
+from dlt.common.schema.utils import get_write_disposition
+
+from dlt.destinations.impl.mssql.configuration import (
+    MsSqlCredentials,
+    MsSqlClientConfiguration,
+)
+from dlt.destinations.impl.mssql.configuration import MsSqlCredentials
+
+
+@configspec
+class SynapseCredentials(MsSqlCredentials):
+    drivername: Final[str] = "synapse"  # type: ignore
+
+    # LongAsMax keyword got introduced in ODBC Driver 18 for SQL Server.
+    SUPPORTED_DRIVERS: ClassVar[List[str]] = ["ODBC Driver 18 for SQL Server"]
+
+    def _get_odbc_dsn_dict(self) -> Dict[str, Any]:
+        params = super()._get_odbc_dsn_dict()
+        # Long types (text, ntext, image) are not supported on Synapse.
+        # Convert to max types using LongAsMax keyword.
+        # https://stackoverflow.com/a/57926224
+        params["LONGASMAX"] = "yes"
+        return params
+
+
+@configspec
+class SynapseClientConfiguration(MsSqlClientConfiguration):
+    destination_type: Final[str] = "synapse"  # type: ignore
+    credentials: SynapseCredentials
+
+    # While Synapse uses CLUSTERED COLUMNSTORE INDEX tables by default, we use
+    # HEAP tables (no indexing) by default. HEAP is a more robust choice, because
+    # columnstore tables do not support varchar(max), nvarchar(max), and varbinary(max).
+    # https://learn.microsoft.com/en-us/azure/synapse-analytics/sql-data-warehouse/sql-data-warehouse-tables-index
+    default_table_index_type: Optional[TTableIndexType] = "heap"
+    """
+    Table index type that is used if no table index type is specified on the resource.
+    This only affects data tables, dlt system tables ignore this setting and
+    are always created as "heap" tables.
+    """
+
+    # Set to False by default because the PRIMARY KEY and UNIQUE constraints
+    # are tricky in Synapse: they are NOT ENFORCED and can lead to innacurate
+    # results if the user does not ensure all column values are unique.
+    # https://learn.microsoft.com/en-us/azure/synapse-analytics/sql-data-warehouse/sql-data-warehouse-table-constraints
+    create_indexes: Optional[bool] = False
+    """Whether `primary_key` and `unique` column hints are applied."""
+
+    # Concurrency is disabled by overriding the configured number of workers to 1 at runtime.
+    auto_disable_concurrency: Optional[bool] = True
+    """Whether concurrency is automatically disabled in cases where it might cause issues."""
+
+    __config_gen_annotations__: ClassVar[List[str]] = [
+        "default_table_index_type",
+        "create_indexes",
+        "auto_disable_concurrency",
+    ]
+
+    def get_load_workers(self, tables: TSchemaTables, workers: int) -> int:
```

Comment:
My take is, that this needs to be an interface for all destinations. So:
* Implement a method on JobClientBase "def enable_parallel_loading" which returns true as a default value
* Evaluate this in the Load class for each destination, and set workers to 1 if it returns false
* Implement this method for synapse, all the other fall back to true.

Reply:
yes, this is good idea. we may need in a few other places (ie. sink destination if user wants to serialize processing).
I'd add it as an option to base configuration of each destination. (should we do it in this PR or a followup?)

Reply:
This is no longer needed, at least not for what I used it for. I replaced the CTAS that caused the concurrency issues with regular `CREATE TABLE` statements. This makes it simpler (no need to manage the number of workers) and also faster because loads can be done concurrently.